File: 210-course-schedule-ii/course-schedule-ii.cpp
class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        
        vector<vector<int>> graph(numCourses);
        vector<int> indeg(numCourses,0);
        for(auto &pre: prerequisites){
            int a =pre[0],b = pre[1];
            graph[b].push_back(a);
            indeg[a]++; 
        }

        queue<int> q;
        for(int i = 0 ;i<numCourses;i++){
            if(indeg[i]==0){
                q.push(i);
            }
        }
        vector<int> order;
        while(!q.empty()){
            int f = q.front();
            q.pop();
            order.push_back(f);
            for(int neighbor: graph[f]){
                indeg[neighbor]--;
                if(indeg[neighbor]==0){
                    q.push(neighbor);
                }
            }
        }
        if(order.size()==numCourses){
            return order;
        }

        else return {};
        
    }
};

File: 210-course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(int numCourses, int[][] prerequisites) {
       List<List<Integer>> res = new ArrayList<>();

       int[] indeg = new int[numCourses];

       for(int i = 0;i<numCourses;i++){
            res.add(new ArrayList<>());
       }

       for(int[] pre : prerequisites){
            int course = pre[0];
            int p = pre[1];

            res.get(p).add(course);
            indeg[course]++;
       }
        Queue<Integer> q = new LinkedList<>();
       for(int i = 0;i<numCourses;i++){
            if(indeg[i] == 0){
                q.add(i);
            }
       }

       int[] result = new int[numCourses];
       int index = 0;

       while(!q.isEmpty()){
            int curr = q.poll();
            result[index++] = curr;

            for(int i : res.get(curr)){
                indeg[i]--;
                if(indeg[i] == 0){
                    q.add(i);
                }
            }
       }

       return index == numCourses ? result : new int[0];
    }
}